servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -249,97 +302,182 @@ public void onComplete() {
+                                final ItemsTerminated<T, B> it = (ItemsTerminated<T, B>) nextState;
+                                terminateIfPossible(it.accumulator, target, it.terminalNotification);
+                            }
+                        }
                         return;
                     }
                 }
             }
         }
 
-        <T, B> void itemsTerminated(final TerminalNotification terminalNotification) {
+        <T, B> void itemsTerminated(final TerminalNotification terminalNotification,
+                                    final Subscriber<? super B> target, final Cancellable bCancellable) {
             for (;;) {
                 final Object cMaybeAccumulator = maybeAccumulator;
+                assert cMaybeAccumulator != ADDING; // `accumulate` and `itemsTerminated` are sequential
                 if (cMaybeAccumulator == TERMINATED) {
                     return;
-                }
-                assert !(cMaybeAccumulator instanceof ItemsTerminated);
-
-                ItemsTerminated<T, B> itemsTerminated = new ItemsTerminated<>(cMaybeAccumulator, terminalNotification);
-                if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, itemsTerminated)) {
-                    return;
+                } else if (cMaybeAccumulator != null &&
+                        NextAccumulatorHolder.class.equals(cMaybeAccumulator.getClass())) {
+                    // Terminated while `nextAccumulator` delivers onNext, notify that thread using ItemsTerminated
+                    // wrapper. There is no more concurrency because neither `itemsTerminated` nor `accumulate` will be
+                    // invoked again.
+                    @SuppressWarnings("unchecked")
+                    final ItemsTerminated<T, B> itemsTerminated = new ItemsTerminated<>(
+                            ((NextAccumulatorHolder<T, B>) cMaybeAccumulator).accumulator, terminalNotification);
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, itemsTerminated)) {
+                        bCancellable.cancel();
+                        return;
+                    }
+                } else if (cMaybeAccumulator == null) {
+                    if (maybeAccumulatorUpdater.compareAndSet(this, null, TERMINATED)) {
+                        // Terminated without any produced items or boundaries
+                        terminateTarget(null, target, terminalNotification);
+                        bCancellable.cancel();
+                        return;
+                    }
+                } else {
+                    @SuppressWarnings("unchecked")
+                    final CountingAccumulator<T, B> accumulator = (CountingAccumulator<T, B>) cMaybeAccumulator;
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator,
+                            new ItemsTerminated<>(accumulator, terminalNotification))) {
+                        bCancellable.cancel();
```

Comment:
above the cancel is done after the downstream termination. can you be consistent, and put this in a `try/finally` (so we always terminate)?

Reply:
There is a case when we cancel in the current thread (line 369), but deliver terminal from another thread that does `nextAccumulator`. I can keep it consistent and always cancel boundaries before termination. Is you suggestion is to move termination to a `finally` block in case `bCancellable.cancel()` throws?

https://github.com/apple/servicetalk/pull/1650/files/d1d008ff9a19a8364e5114efcbcbf8edc79576f7#diff-a02114146aee33b0d470d21bdd07e3e9017f55b6aea426a264ee4c8d62a6efaaR360-R370

Reply:
I made cancellation consistent and always cancel boundaries before propagating termination. Also added `try-finally` block to make sure the terminal event is propagated.